lightning/src/ln/channelmanager.rs

```diff
@@ -1786,6 +1794,25 @@ where
 		self.list_channels_with_filter(|&(_, ref channel)| channel.is_live())
 	}
 
+	/// Gets the list of channels we have with a given counterparty, in random order.
+	pub fn list_channels_with_counterparty(&self, counterparty_node_id: &PublicKey) -> Vec<ChannelDetails> {
+		let mut res = Vec::with_capacity(self.short_to_chan_info.read().unwrap().len());
```

Comment:
Won't this allocate too much? You should use `peer_state.channel_by_id.len()` instead.

Also, I *think* if you use `iter().map(...).collect()`, it's smart enough to allocated enough capacity.

```diff
diff --git a/lightning/src/ln/channelmanager.rs b/lightning/src/ln/channelmanager.rs
index 6562d30a..042ff9d2 100644
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -1796,21 +1796,19 @@ where
 
        /// Gets the list of channels we have with a given counterparty, in random order.
        pub fn list_channels_with_counterparty(&self, counterparty_node_id: &PublicKey) -> Vec<ChannelDetails> {
-               let mut res = Vec::with_capacity(self.short_to_chan_info.read().unwrap().len());
-               {
-                       let best_block_height = self.best_block.read().unwrap().height();
-                       let per_peer_state = self.per_peer_state.read().unwrap();
-                       if let Some(peer_state_mutex) = per_peer_state.get(counterparty_node_id) {
-                               let mut peer_state_lock = peer_state_mutex.lock().unwrap();
-                               let peer_state = &mut *peer_state_lock;
-                               for (_channel_id, channel) in peer_state.channel_by_id.iter() {
-                                       let details = ChannelDetails::from_channel(channel, best_block_height,
-                                               peer_state.latest_features.clone());
-                                       res.push(details);
-                               }
-                       }
-               }
-               res
+               let best_block_height = self.best_block.read().unwrap().height();
+               let per_peer_state = self.per_peer_state.read().unwrap();
+               if let Some(peer_state_mutex) = per_peer_state.get(counterparty_node_id) {
+                       let mut peer_state_lock = peer_state_mutex.lock().unwrap();
+                       let peer_state = &mut *peer_state_lock;
+                       let features = &peer_state.latest_features;
+                       return peer_state.channel_by_id
+                               .iter()
+                               .map(|(_, channel)|
+                                       ChannelDetails::from_channel(channel, best_block_height, features.clone()))
+                               .collect();
+               }
+               vec![]
        }
 
        /// Returns in an undefined order recent payments that -- if not fulfilled -- have yet to find a
```